[BUG] Duplicated API calls from frontend on text filtering

I have now over 400 documents in my paperless installation. Nearly 100 documents of them are fairly large (200 pages color).

Paperless-ngx V1.7.1 with postgres

When typing ~~slowly~~ at normal speed (not Superman fast...) some long search keyword into the "title and content" field at "documents" it seems that there is for every letter a new SELECT statement transferred to the postgres instance. This leads to a "flickering" effect of matching thumbnails/documents. It shows the new result for half a second, jumps back to the origin view, show another new thumbnail preview, jumps back to the origin view and so on. It takes couple of seconds when finally the final result view for the actual search keyword is found and "flickering" ends.

#############
It already does have a debounce, thats not really the same issue.

_Originally posted by @shamoon in https://github.com/paperless-ngx/paperless-ngx/discussions/288#discussioncomment-2354915_

#############

Your fix works great, only one request is made now. No more flickering in the preview! Thank you. (Search starts after the 3rd letter is entered as expected, this is for 1.7.1 and dev build for me. So no problem with that.)

But still I would say the debounce timer for a key press is too short. How many ms are now set?

Some other observations with this dev build: The conversion of thumbnails to webp was smooth and fast (414 images in nearly 6 seconds (on a 6 core Intel i5 with only one paperless worker and SATA SSDs).

Some error message I got with this build:

```
10:47:15 [Q] INFO recycled worker Process-1:1
10:47:15 [Q] INFO Process-1:6 ready for work at 495
[2022-06-15 10:48:45 +0200] [470] [CRITICAL] WORKER TIMEOUT (pid:473)
[2022-06-15 10:48:45 +0200] [470] [WARNING] Worker with pid 473 was terminated due to signal 6
10:51:40 [Q] INFO Enqueued 1
```
```
[2022-06-15 10:51:41,027] [ERROR] [paperless.classifier] Unrecoverable error while loading document classification model, deleting model file.
Traceback (most recent call last):
  File "/usr/src/paperless/src/documents/classifier.py", line 40, in load_classifier
    classifier.load()
  File "/usr/src/paperless/src/documents/classifier.py", line 82, in load
    raise IncompatibleClassifierVersionError(
documents.classifier.IncompatibleClassifierVersionError: Cannot load classifier, incompatible versions.
```
Postgres:

```
2022-06-15 08:46:36.585 UTC [34] ERROR:  column documents_document.storage_path_id does not exist at character 76
2022-06-15 08:46:36.585 UTC [34] STATEMENT:  SELECT "documents_document"."id", "documents_document"."correspondent_id", "documents_document"."storage_path_id", "documents_document"."title", "documents_document"."document_type_id", "documents_document"."content", "documents_document"."mime_type", "documents_document"."checksum", "documents_document"."archive_checksum", "documents_document"."created", "documents_document"."modified", "documents_document"."storage_type", "documents_document"."added", "documents_document"."filename", "documents_document"."archive_filename", "documents_document"."archive_serial_number" FROM "documents_document" WHERE "documents_document"."storage_type" = 'gpg' ORDER BY "documents_document"."created" DESC LIMIT 1
```

--- a/src-ui/src/app/components/document-list/filter-editor/filter-editor.component.ts
+++ b/src-ui/src/app/components/document-list/filter-editor/filter-editor.component.ts
@@ -16,7 +16,7 @@ import { debounceTime, distinctUntilChanged, filter } from 'rxjs/operators'
 import { DocumentTypeService } from 'src/app/services/rest/document-type.service'
 import { TagService } from 'src/app/services/rest/tag.service'
 import { CorrespondentService } from 'src/app/services/rest/correspondent.service'
-import { FilterRule } from 'src/app/data/filter-rule'
+import { filterRulesDiffer, FilterRule } from 'src/app/data/filter-rule'
 import {
   FILTER_ADDED_AFTER,
   FILTER_ADDED_BEFORE,
@@ -204,7 +204,10 @@ export class FilterEditorComponent implements OnInit, OnDestroy {
   @Input()
   set unmodifiedFilterRules(value: FilterRule[]) {
     this._unmodifiedFilterRules = value
-    this.checkIfRulesHaveChanged()
+    this.rulesModified = filterRulesDiffer(
+      this._unmodifiedFilterRules,
+      this._filterRules
+    )
   }
 
   get unmodifiedFilterRules(): FilterRule[] {
@@ -330,7 +333,10 @@ export class FilterEditorComponent implements OnInit, OnDestroy {
           break
       }
     })
-    this.checkIfRulesHaveChanged()
+    this.rulesModified = filterRulesDiffer(
+      this._unmodifiedFilterRules,
+      this._filterRules
+    )
   }
 
   get filterRules(): FilterRule[] {
@@ -472,31 +478,6 @@ export class FilterEditorComponent implements OnInit, OnDestroy {
   filterRulesChange = new EventEmitter<FilterRule[]>()
 
   rulesModified: boolean = false
-
-  private checkIfRulesHaveChanged() {
-    let modified = false
-    if (this._unmodifiedFilterRules.length != this._filterRules.length) {
-      modified = true
-    } else {
-      modified = this._unmodifiedFilterRules.some((rule) => {
-        return (
-          this._filterRules.find(
-            (fri) => fri.rule_type == rule.rule_type && fri.value == rule.value
-          ) == undefined
-        )
-      })
-
-      if (!modified) {
-        // only check other direction if we havent already determined is modified
-        modified = this._filterRules.some((rule) => {
-          this._unmodifiedFilterRules.find(
-            (fr) => fr.rule_type == rule.rule_type && fr.value == rule.value
-          ) == undefined
-        })
-      }
-    }
-    this.rulesModified = modified
-  }
 
   updateRules() {
     this.filterRulesChange.next(this.filterRules)

--- a/src-ui/src/app/data/filter-rule.ts
+++ b/src-ui/src/app/data/filter-rule.ts
@@ -25,6 +25,34 @@ export function isFullTextFilterRule(filterRules: FilterRule[]): boolean {
   )
 }
 
+export function filterRulesDiffer(
+  filterRulesA: FilterRule[],
+  filterRulesB: FilterRule[]
+): boolean {
+  let modified = false
+  if (filterRulesA.length != filterRulesB.length) {
+    modified = true
+  } else {
+    modified = filterRulesA.some((rule) => {
+      return (
+        filterRulesB.find(
+          (fri) => fri.rule_type == rule.rule_type && fri.value == rule.value
+        ) == undefined
+      )
+    })
+
+    if (!modified) {
+      // only check other direction if we havent already determined is modified
+      modified = filterRulesB.some((rule) => {
+        filterRulesA.find(
+          (fr) => fr.rule_type == rule.rule_type && fr.value == rule.value
+        ) == undefined
+      })
+    }
+  }
+  return modified
+}
+
 export interface FilterRule {
   rule_type: number
   value: string

--- a/src-ui/src/app/data/filter-rule.ts
+++ b/src-ui/src/app/data/filter-rule.ts
@@ -29,28 +29,19 @@ export function filterRulesDiffer(
   filterRulesA: FilterRule[],
   filterRulesB: FilterRule[]
 ): boolean {
-  let modified = false
+  let differ = false
   if (filterRulesA.length != filterRulesB.length) {
-    modified = true
+    differ = true
   } else {
-    modified = filterRulesA.some((rule) => {
+    differ = filterRulesA.some((rule) => {
       return (
         filterRulesB.find(
           (fri) => fri.rule_type == rule.rule_type && fri.value == rule.value
         ) == undefined
       )
     })
-
-    if (!modified) {
-      // only check other direction if we havent already determined is modified
-      modified = filterRulesB.some((rule) => {
-        filterRulesA.find(
-          (fr) => fr.rule_type == rule.rule_type && fr.value == rule.value
-        ) == undefined
-      })
-    }
   }
-  return modified
+  return differ
 }
 
 export interface FilterRule {

--- a/src-ui/src/app/services/document-list-view.service.ts
+++ b/src-ui/src/app/services/document-list-view.service.ts
@@ -2,6 +2,7 @@ import { Injectable } from '@angular/core'
 import { ParamMap, Router } from '@angular/router'
 import { Observable } from 'rxjs'
 import {
+  filterRulesDiffer,
   cloneFilterRules,
   FilterRule,
   isFullTextFilterRule,
@@ -67,6 +68,7 @@ export interface ListViewState {
 })
 export class DocumentListViewService {
   isReloading: boolean = false
+  initialized: boolean = false
   error: string = null
 
   rangeSelectionAnchorIndex: number
@@ -172,11 +174,24 @@ export class DocumentListViewService {
     if (!paramsEmpty) newState = parseParams(queryParams)
     if (newState == undefined) newState = this.defaultListViewState() // if nothing in local storage
 
-    this.activeListViewState.filterRules = newState.filterRules
-    this.activeListViewState.sortField = newState.sortField
-    this.activeListViewState.sortReverse = newState.sortReverse
-    this.activeListViewState.currentPage = newState.currentPage
-    this.reload(null, paramsEmpty) // update the params if there arent any
+    // only reload if things have changed
+    if (
+      !this.initialized ||
+      paramsEmpty ||
+      this.activeListViewState.sortField !== newState.sortField ||
+      this.activeListViewState.sortReverse !== newState.sortReverse ||
+      this.activeListViewState.currentPage !== newState.currentPage ||
+      filterRulesDiffer(
+        this.activeListViewState.filterRules,
+        newState.filterRules
+      )
+    ) {
+      this.activeListViewState.filterRules = newState.filterRules
+      this.activeListViewState.sortField = newState.sortField
+      this.activeListViewState.sortReverse = newState.sortReverse
+      this.activeListViewState.currentPage = newState.currentPage
+      this.reload(null, paramsEmpty) // update the params if there arent any
+    }
   }
 
   reload(onFinish?, updateQueryParams: boolean = true) {
@@ -193,6 +208,7 @@ export class DocumentListViewService {
       )
       .subscribe({
         next: (result) => {
+          this.initialized = true
           this.isReloading = false
           activeListViewState.collectionSize = result.count
           activeListViewState.documents = result.results

